File: app/src/main/java/com/cn29/aac/ui/main/AppArchNavigationDrawer.java
package com.cn29.aac.ui.main;

import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;

import com.cn29.aac.R;
import com.cn29.aac.databinding.ActivityAppArchNavigationDrawerBinding;
import com.cn29.aac.ui.base.BaseAppCompatActivity;
import com.cn29.aac.ui.feedentry.FeedActivity;
import com.cn29.aac.ui.location.LocationActivity;
import com.cn29.aac.ui.login.LoginActivity;
import com.cn29.aac.ui.main.vm.AppArchNavViewModel;
import com.cn29.aac.ui.masterdetail.SimpleListActivity;
import com.cn29.aac.ui.setting.SettingsActivity;
import com.cn29.aac.ui.shopping.ShoppingActivity;
import com.cn29.aac.ui.shoppingkart.ShoppingKartActivity;
import com.cn29.aac.ui.viewpager.PagerActivity;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.navigation.NavigationView;
import com.google.android.material.snackbar.Snackbar;

import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class AppArchNavigationDrawer extends BaseAppCompatActivity
    implements NavigationView.OnNavigationItemSelectedListener {

  @Inject
  AppArchNavViewModel appArchNavViewModel;

  @Inject
  ActivityAppArchNavigationDrawerBinding binding;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    Toolbar toolbar = findViewById(R.id.toolbar);
    setSupportActionBar(toolbar);
    FloatingActionButton fab = findViewById(R.id.fab);
    fab.setOnClickListener(
        view -> Snackbar.make(view, "Want to set email ?", Snackbar.LENGTH_LONG)
            .setAction("Action", null).show());
    DrawerLayout drawer = findViewById(R.id.drawer_layout);
    ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
        this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
    drawer.addDrawerListener(toggle);
    toggle.syncState();
    NavigationView navigationView = findViewById(R.id.nav_view);
    navigationView.setNavigationItemSelectedListener(this);
  }

  @Override
  public void onBackPressed() {
    DrawerLayout drawer = findViewById(R.id.drawer_layout);
    if (drawer.isDrawerOpen(GravityCompat.START)) {
      drawer.closeDrawer(GravityCompat.START);
    } else {
      super.onBackPressed();
    }
  }


  @SuppressWarnings("StatementWithEmptyBody")
  @Override
  public boolean onNavigationItemSelected(@NonNull MenuItem item) {
    // Handle navigation view item clicks here.
    int id = item.getItemId();
    switch (id) {
      case R.id.nav_fragment_communication:
        startActivity(new Intent(this, SimpleListActivity.class));
        break;
      case R.id.nav_storage:
        startActivity(new Intent(this, FeedActivity.class));
        break;
      case R.id.nav_connectivity:
        startActivity(new Intent(this, PagerActivity.class));
        break;
      case R.id.nav_location:
        startActivity(new Intent(this, LocationActivity.class));
        break;
      case R.id.nav_shopping:
        startActivity(new Intent(this, ShoppingActivity.class));
        break;
      case R.id.nav_shopping_kart:
        startActivity(new Intent(this, ShoppingKartActivity.class));
        break;
      case R.id.nav_shopping_history:
        viewHistory();
        break;
      case R.id.nav_ordering:
        ordering();
        break;
      case R.id.nav_settings:
        openSettings();
        break;
      case R.id.nav_logout:
        logout();
        break;
    }
    DrawerLayout drawer = findViewById(R.id.drawer_layout);
    drawer.closeDrawer(GravityCompat.START);
    return true;
  }

  private void openSettings() {
    startActivity(new Intent(this, SettingsActivity.class));
  }

  private void viewHistory() {

  }

  private void ordering() {
  }

  private void logout() {
    progressDialogComponent.showLoading();
    //clear the data first
    appArchNavViewModel.logout();
    Single.timer(3, TimeUnit.SECONDS)
        .subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(s -> {
          //finish the current activity and go back to the login activity
          progressDialogComponent.hideLoading();
          finish();
          startActivity(new Intent(this, LoginActivity.class));
        });
  }
}

File: app/src/main/java/com/cn29/aac/ui/main/AppArchNavigationDrawer.kt
package com.cn29.aac.ui.main

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.widget.Toolbar
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import com.cn29.aac.R
import com.cn29.aac.databinding.ActivityAppArchNavigationDrawerBinding
import com.cn29.aac.ui.base.BaseAppCompatActivity
import com.cn29.aac.ui.feedentry.FeedActivity
import com.cn29.aac.ui.location.LocationActivity
import com.cn29.aac.ui.login.LoginActivity
import com.cn29.aac.ui.main.vm.AppArchNavViewModel
import com.cn29.aac.ui.masterdetail.SimpleListActivity
import com.cn29.aac.ui.setting.SettingsActivity
import com.cn29.aac.ui.shopping.ShoppingActivity
import com.cn29.aac.ui.shoppingkart.ShoppingKartActivity
import com.cn29.aac.ui.viewpager.PagerActivity
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.navigation.NavigationView
import com.google.android.material.snackbar.Snackbar
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class AppArchNavigationDrawer : BaseAppCompatActivity(),
        NavigationView.OnNavigationItemSelectedListener {
    @JvmField
    @Inject
    var appArchNavViewModel: AppArchNavViewModel? = null

    @JvmField
    @Inject
    var binding: ActivityAppArchNavigationDrawerBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val toolbar = findViewById<Toolbar>(
                R.id.toolbar)
        setSupportActionBar(toolbar)
        val fab = findViewById<FloatingActionButton>(R.id.fab)
        fab.setOnClickListener { view: View? ->
            Snackbar.make(view!!,
                          "Want to set email ?",
                          Snackbar.LENGTH_LONG)
                    .setAction("Action",
                               null)
                    .show()
        }
        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        val toggle = ActionBarDrawerToggle(
                this,
                drawer,
                toolbar,
                R.string.navigation_drawer_open,
                R.string.navigation_drawer_close)
        drawer.addDrawerListener(toggle)
        toggle.syncState()
        val navigationView = findViewById<NavigationView>(R.id.nav_view)
        navigationView.setNavigationItemSelectedListener(this)
    }

    override fun onBackPressed() {
        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        val id = item.itemId
        when (id) {
            R.id.nav_fragment_communication -> startActivity(Intent(this,
                                                                    SimpleListActivity::class.java))
            R.id.nav_storage -> startActivity(Intent(this,
                                                     FeedActivity::class.java))
            R.id.nav_connectivity -> startActivity(Intent(this,
                                                          PagerActivity::class.java))
            R.id.nav_location -> startActivity(Intent(this,
                                                      LocationActivity::class.java))
            R.id.nav_shopping -> startActivity(Intent(this,
                                                      ShoppingActivity::class.java))
            R.id.nav_shopping_kart -> startActivity(Intent(this,
                                                           ShoppingKartActivity::class.java))
            R.id.nav_shopping_history -> viewHistory()
            R.id.nav_ordering -> ordering()
            R.id.nav_settings -> openSettings()
            R.id.nav_logout -> logout()
        }
        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        drawer.closeDrawer(GravityCompat.START)
        return true
    }

    private fun openSettings() {
        startActivity(Intent(this, SettingsActivity::class.java))
    }

    private fun viewHistory() {}
    private fun ordering() {}
    private fun logout() {
        progressDialogComponent!!.showLoading()
        //clear the data first
        appArchNavViewModel!!.logout()
        val subscribe = Single.timer(3, TimeUnit.SECONDS)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { _: Long? ->
                    //finish the current activity and go back to the login activity
                    progressDialogComponent!!.hideLoading()
                    finish()
                    startActivity(
                            Intent(this,
                                   LoginActivity::class.java))
                }
    }
}